src/mpid/ch4/shm/xpmem/xpmem_init.c

```diff
@@ -68,6 +88,20 @@ int MPIDI_XPMEM_mpi_init_hook(int rank, int size, int *n_vcis_provided, int *tag
     /* Initialize other global parameters */
     MPIDI_XPMEM_global.sys_page_sz = (size_t) sysconf(_SC_PAGESIZE);
 
+    /* Attach to local root coop_counter array */
+    if (MPIR_Process.local_rank) {
+        mpi_errno =
+            MPIDI_XPMEM_seg_regist(0, num_local * num_local * sizeof(OPA_int_t),
+                                   (void *) coop_caddr, &seg_ptr, (void **) &coop_counter);
+        if (mpi_errno != MPI_SUCCESS)
+            MPIR_ERR_POP(mpi_errno);
+    }
```

Comment:
- The `coop_counter` is only internally used. Can we `xpmem_attach` it instead of registering into the AVL tree? Otherwise it increases searching overhead when we register every user buffer, right?

- Can you replace `seg_ptr` with a better name? Maybe you do not need this attribute any more if just attach it.

- Maybe it is better to allocate page aligned size for `coop_counter`. It ensures the same page is never shared by user buffer.

- I'd suggest that you allocate `coop_counter` and exchange it after `anyfail` check. And please define the local variables near the place where it is used (we decided to follow the `c99` code style in MPICH)

Reply:
- Each process only adds one segment into AVL that does not cost much searching overhead. I think it's ok to keep it there.
- Please consider my first point.
- Good point, I will change it to page aligned one.
- Are you mentioning 'num_local' variable in `xpmem_init.c`? I found sometimes variable is defined near the place where it is used, but most of the time variables are defined in the beginning of routines. Which standard should I follow?